src/drivers/roboclaw/RoboClaw.cpp

```diff
@@ -112,84 +135,147 @@ RoboClaw::~RoboClaw()
 	close(_uart);
 }
 
-int RoboClaw::readEncoder(e_motor motor)
+void RoboClaw::taskMain()
 {
-	uint16_t sum = 0;
 
-	if (motor == MOTOR_1) {
-		_sendCommand(CMD_READ_ENCODER_1, nullptr, 0, sum);
+	// This main loop performs two different tasks, asynchronously:
+	// - Send actuator_controls_0 to the Roboclaw as soon as they are available
+	// - Read the encoder values at a constant rate
+	// To do this, the timeout on the poll() function is used.
+	// waitTime is the amount of time left (int microseconds) until the next time I should read from the encoders.
+	// It is updated at the end of every loop. Sometimes, if the actuator_controls_0 message came in right before
+	// I should have read the encoders, waitTime will be 0. This is fine. When waitTime is 0, poll() will return
+	// immediately with a timeout. (Or possibly with a message, if one happened to be available at that exact moment)
+	uint64_t encoderTaskLastRun = 0;
+	int waitTime = 0;
 
-	} else if (motor == MOTOR_2) {
-		_sendCommand(CMD_READ_ENCODER_2, nullptr, 0, sum);
-	}
+	_actuatorsSub = orb_subscribe(ORB_ID(actuator_controls_0));
+	orb_set_interval(_actuatorsSub, ACTUATOR_WRITE_PERIOD_MS);
 
-	uint8_t rbuf[50];
-	usleep(5000);
-	int nread = read(_uart, rbuf, 50);
+	_armedSub = orb_subscribe(ORB_ID(actuator_armed));
 
-	if (nread < 6) {
-		printf("failed to read\n");
-		return -1;
-	}
+	//TODO: Add parameter listening
+	pollfd fds[2];
+	fds[0].fd = _actuatorsSub;
+	fds[0].events = POLLIN;
+	fds[1].fd = _armedSub;
+	fds[1].events = POLLIN;
 
-	//printf("received: \n");
-	//for (int i=0;i<nread;i++) {
-	//printf("%d\t", rbuf[i]);
-	//}
-	//printf("\n");
-	uint32_t count = 0;
-	uint8_t *countBytes = (uint8_t *)(&count);
-	countBytes[3] = rbuf[0];
-	countBytes[2] = rbuf[1];
-	countBytes[1] = rbuf[2];
-	countBytes[0] = rbuf[3];
-	uint8_t status = rbuf[4];
-	uint8_t checksum = rbuf[5];
-	uint8_t checksum_computed = (sum + _sumBytes(rbuf, 5)) &
-				    checksum_mask;
-
-	// check if checksum is valid
-	if (checksum != checksum_computed) {
-		printf("checksum failed: expected %d got %d\n",
-		       checksum, checksum_computed);
-		return -1;
-	}
+	memset((void *) &_wheelEncoderMsg, 0, sizeof(wheel_encoders_s));
+	_wheelEncoderMsg.timestamp = hrt_absolute_time();
+	_wheelEncodersAdv = orb_advertise(ORB_ID(wheel_encoders), &_wheelEncoderMsg);
+
+	while (!taskShouldExit) {
 
-	int overFlow = 0;
+		int pret = poll(fds, sizeof(fds) / sizeof(pollfd), waitTime / 1000);
 
-	if (status & STATUS_REVERSE) {
-		//printf("roboclaw: reverse\n");
+		// No timeout, update on either the actuator controls or the armed state
+		if (pret > 0 && (fds[0].revents & POLLIN || fds[1].revents & POLLIN)) {
+			orb_copy(ORB_ID(actuator_controls_0), _actuatorsSub, &_actuatorControls);
+			orb_copy(ORB_ID(actuator_armed), _armedSub, &_actuatorArmed);
+
+			int drive_ret = 0, turn_ret = 0;
+
+			// Disarmed
```

Comment:
Instead of a comment you could do:
```
const bool disarmed = (!_actuatorArmed.armed || _actuatorArmed.lockdown || _actuatorArmed.manual_lockdown
			    || _actuatorArmed.force_failsafe);
if (disarmed) {
    ...
```